### PR TITLE
HTTP Resource Representation Variants

### PR DESCRIPTION
**Pull request by snuggs**
This was already implemented but was sporadically problematic with indeterministic bugs. In other words...duct taped together with **no tests**. Tests now thanks to tap(e).

## Installation
Ability to "install" snuggsi packaged library from `https://snuggsi.es` with the following:
```html
<script src=//snuggsi.es></script>
```

Or from mounting  server middleware

#### server.js
```javascript
app // Express or Koa
  .use ( snuggsi )
```

#### index.html
```html
<!-- served from `/snuggsi` endpoint (minified and compressed) -->
<!-- (Javascript / Ecmascript negotiated on browser capabilities) -->
<!-- (gzip / brotli negotiated on browser capabilities) -->
<script src=/snuggsi></script>

<!-- served from `/snuggsi` endpoint (minified and compressed) -->
<!-- (Javascript / Ecmascript negotiated on browser capabilities) -->
<script src=/snuggsi?debug></script>

<!-- Ecmascript version (minified and compressed) -->
<!-- (gzip / brotli negotiated on browser capabilities) -->
<script src=/snuggsi.es></script>

<!-- Ecmascript version (uncompressed) -->
<script src=/snuggsi.es?debug></script>

<!-- JavaScript version (minified and compressed) -->
<!-- (gzip / brotli negotiated on browser capabilities) -->
<script src=/snuggsi.js></script>

<!-- JavaScript version (uncompressed) -->
<script src=/snuggsi.js?debug></script>
```

##  References:
- HTTP Representation Variants - https://mnot.github.io/I-D/variants
- WHATWG URL Spec - https://url.spec.whatwg.org
- Wikipedia - https://en.wikipedia.org/wiki/Content_negotiation
- MDN `Accept-Encoding` - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding
- MDN - https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation
- HTTP Accept Header - https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.1

### Resource Default HTML `GET / Accept: text/html`

- [x] - `GET /?debug Accept: text/html` - _(`index.html`)_ with no compression.
- [x] - `GET / Accept: text/html Accept-Encoding: identity,(none)` - _(`index.html`)_ with no compression.

### Resource Default CSS `GET / Accept: text/css`

- [x] - `GET /?debug Accept: text/css` - _(`index.css')_ with no compression.
- [x] - `GET / Accept: text/css Accept-Encoding: identity,(none)` - _(`index.css`)_ no compression.

### Resource Default Javascript `GET / Accept: */*`

**Will determine whether or not to serve modern Ecmascript or legacy JavaScript**

- [x] - `GET /?debug Accept: */*` - `application/javascript` _(`.js`)_ no compression.
- [x] - `GET /?debug Accept: */* Accept-Encoding: br` - `application/ecmascript` _(`.es`)_ no compression.
- [x] - `GET / Accept: */* Accept-Encoding: identity,(none)` - `application/javascript` _(`.js`)_ no compression.
- [x] - `GET / Accept: */* Accept-Encoding: br` - `application/ecmascript` _(`.min.es.br`)_ with brotli compression.
- [x] - `GET / Accept: */* Accept-Encoding: gzip,deflate` - `application/javascript` _(`.min.js.gz`)_ gzip compression.

### `GET /snuggsi`

**Will determine whether or not to serve modern Ecmascript or legacy JavaScript**

- [x] - `GET /snuggsi?debug` - `application/ecmascript` _(`.js`)_ with no compression.
- [x] - `GET /snuggsi?debug Accept-Encoding: br,identity` - `application/ecmascript` _(`.es`)_ with no compression.
- [x] - `GET /snuggsi?debug Accept-Encoding: gzip,compress,deflate,identity,*,(none)` - `application/javascript` _(`.js`)_ with no compression.
- [x] - `GET /snuggsi - `application/ecmascript` _(`.js`)_ with no compression.
- [x] - `GET /snuggsi Accept-Encoding: identity` - `application/javascript` _(`.js`)_ with no compression.
- [x] - `GET /snuggsi Accept-Encoding: *` - `application/javascript` _(`.min.js.gz`)_ with gzip compression.
- [x] - `GET /snuggsi Accept-Encoding: gzip` - `application/javascript` _(`.min.js.gz`)_ with gzip compression.
- [x] - `GET /snuggsi Accept-Encoding: br` - `application/ecmascript` _(`.min.es.br`)_ with brotli compression.

### `GET /snuggsi.es`

- [x] - `GET /snuggsi.es` - `application/ecmascript` with no compression.
- [x] - `GET /snuggsi.es?debug` - `application/ecmascript` with no compression.
- [x] - `GET /snuggsi.es Accept-Encoding: identity` - `application/javascript` with no compression.
- [x] - `GET /snuggsi.es Accept-Encoding: *` - `application/ecmascript` with default compression.
- [x] - `GET /snuggsi.es Accept-Encoding: gzip` - `application/ecmascript` with gzip compression.
- [x] - `GET /snuggsi.es Accept-Encoding: br` - `application/ecmascript` with brotli compression.

### `GET /snuggsi.js`

- [x] - `GET /snuggsi.js` - `application/javascript` with no compression.
- [x] - `GET /snuggsi.js?debug` - `application/javascript` with no compression.
- [x] - `GET /snuggsi.js Accept-Encoding: identity` - `application/javascript` with no compression.
- [x] - `GET /snuggsi.js Accept-Encoding: *` - `application/javascript` with default compression.
- [x] - `GET /snuggsi.js Accept-Encoding: gzip` `application/javascript` with gzip compression.
- [x] - `GET /snuggsi.js Accept-Encoding: br` - `application/javascript` with brotli compression.

## References
  - [WHATWG Mime sniffing](https://mimesniff.spec.whatwg.org/#javascript-mime-type)
  - [Web compatible javascript media types](https://html.spec.whatwg.org/multipage/scripting.html#javascript-mime-type)
  - https://rubyplus.com/articles/3231-How-To-Get-A-List-Of-All-Currently-Registered-MIME-Types-With-Rails
  - [Javascript Media Types on the web. The saga continues](https://github.com/nodejs/node-eps/pull/61#issuecomment-321329296)
  - https://github.com/whatwg/html/issues/558
  - [WHATWG `fetch`ing a single module algorithm](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script)

## ES6 Modules & Client/Server support

Anyone touching an app in the browser should at least get familiar.
/cc @mrbernnz @btakita @halorium @janz93 @pachonk @robcole @scottmacdowell @kurtcagle @johnrlive
  - http://jsmodules.io
  - https://blog.whatwg.org/js-modules
  - http://2ality.com/2017/05/es-module-specifiers.html
  - (@tmornini  you thought THIS PR was Epic. lol) https://github.com/whatwg/html/pull/443

### CODE REVIEW

**Comment by tmornini** [comment]
@snuggs WOW! Huge piece of work!

**Comment by snuggs** [comment]
@tmornini indeed! And VERY well tested :-)

**Comment by mrbernnz** [comment]
@snuggs did this PR really need 250+ commits?

**Comment by tmornini** [comment]
@mrbernnz Strange questions to ask when staring at a 250+ commit PR. 😄

Of course, @snuggs may well decide to do some rebasing before going to master.

**Comment by brandondees** [comment]
what is this, github golf? ⛳️

**Comment by tmornini** [comment]
@brandondees If so, we're going to have to start calling @snuggs "Tiger" 😄

**Comment by snuggs** [comment]
@mrbernnz there are over 55 different permutations for matching routes to endpoints with HTTP verbs. Let alone content negotiation for the types. Imagine that truth table. Also working with MDN, WHATWG, and W3C on doing SOUND HTTP Documentation for the web. Let's just say it doesn't move as fast. 

Just be readin' up on the description links about HTTP variants and resources....That is unless you want to slang commits too! ;-)

You will see that (finally) we are moving in a direction towards HTTP that @tmornini and I been talking about for damn near a decade. HTTP > MVC. [Flame is based on many of the same principles of this PR](https://github.com/AlexWayfer/flame/pull/37)

The intent of this PR is because i'm sick of having a "God directory" for assets and endpoints when that's not how HTTP works.

```
class {
  // You get get, head, options for free because the HTTP spec says so
  options (context)
    { context.status = 204; context.body = 'Or you could monkeypatch but not advisable' }
  post (context) {}
  put (context) {}

// @tmornini get / purge is where caching is done YES?! ;-) 
}
```
Amazes me we took this long to realize there are no Controllers....only resources... 

To be clear this is for internal use only but that won't be the case by year's end.

P.S. [There's a test for every single one of those permutations as well 😎  ](https://github.com/devpunks/snuggsi/blob/4024e2cb33443f26a7b27fc46c4ee9a7508f5ee6/resource/index.test). /cc @tmornini @mrbernnz.

**Comment by snuggs** [comment]
@tmornini @mrbernnz @brandondees interesting...
Feeling even better about this PR.

How many frameworks you think would go 💥 💣 if you went back in retrospect and implemented the HTTP spec correctly....

https://github.com/gin-gonic/gin/issues/20
https://github.com/julienschmidt/httprouter/pull/51
https://github.com/julienschmidt/httprouter/issues/30
https://github.com/julienschmidt/httprouter/issues/52

**Comment by brandondees** [comment]
@snuggs hard to say i could expect any of them to work with it fully without taking it all back to the drawing board. Looks like the example cited is the closest thing I know of, and still doesn't have all the kinks worked out yet.

Imagine if you could run a _standard suite of tests_ that could assert spec compliance for a given implementation...

**Comment by snuggs** [comment]
> Imagine if you could run a standard suite of tests that could assert spec compliance for a given implementation... @brandondees 

Interesting you say this. About a year ago @tmornini and I had a discussion about versions/testing living specs. I mentioned to him how even I don't know if the suite passes and that's great we have CI. The specs change frequently and implementation can break just from an update to [WPT (Web Platform Tests. The test suite that ALL browsers MUST adhere to.)](https://github.com/w3c/web-platform-tests) And we use [JSDOM](https://github.com/jsdom/jsdom) which basically creates an in memory "browser" that is merely a stubbed implementation of said WPT/WHATWG/W3C specs. 

Browser vendors are constantly breaking tests (that were previously passing). An update to specs (therefore update to WPT) triggers failing tests and tell the browsers what they MUST implement immediately. None of the spec's concern either....an implementation detail. Personal problem of the browser vendors if you will. We are also under the same scrutiny which I think is a great thing.

That being said, to make a long story longer your thoughts may not be too out of reach in 2018. We're definitely on that path already (intentionally).

**Comment by snuggs** [comment]
@brandondees @tmornini PERFECT example of the spec being changed (which now breaks our implementation). Luckily we didn't expose `.upgrade (node)` but internally we will need to implement this algorithm. @brandondees feel like making a bug?

Issue Request from W3C: https://github.com/w3c/webcomponents/issues/710
PR Response from WHATWG: https://github.com/whatwg/html/pull/3535
LOC in question _(@tmornini  when you're TOO good at #CATAT Call A Thing A Thing)_: https://github.com/devpunks/snuggsi/blame/master/custom-element-registry/index.es#L32
Tests to pass: https://github.com/w3c/web-platform-tests/pull/9869
Browser Bugs: https://github.com/whatwg/html/pull/3535#issuecomment-371027490

We got time before the browsers get there LOLZ but doesn't mean we can't implement today. 

P.S. @brandondees still steering CLEAR AWAY from Shadow DOMmy stuff. YAGNI (yet). We support `<slot>` tho. Please respond with an issue not comment on unrelated PR that's becomming long in the tooth if not merged this week hehe. 

/cc @tmornini

**Comment by snuggs** [comment]
@brandondees we'll need to update Rails, Sinatra, Express, Koa for the new `.mjs` and existing `.es` extensions. 

I'm on it this weekend. This list and `Resource.assets` will be just about it for this PR. And specs are fairly in order as media types have been an ancient 💩 show for about a decade now.

## Related

  - [x] [Submit proposal for `.mjs` to IETF](https://datatracker.ietf.org/submit/status/88872)
  - [x] [Un-deprecate `text/javascript` Media type](https://datatracker.ietf.org/submit/status/88872)
  - [x] [Update `application/ecmascript` to respect `.mjs` extension](https://datatracker.ietf.org/submit/status/88872)
  - [x] [Update to reflect `application/ecmascript`](https://github.com/jshttp/mime-db/pull/116)
  - [x] [Add `.mjs` to mimetypes database](https://github.com/jshttp/mime-db/pull/88)
  - [ ] Add `.mjs` to mime types database with [The following WHATWG approved Media Types](https://html.spec.whatwg.org/multipage/scripting.html#javascript-mime-type)
  - [ ] 👊 bump `koa` https://github.com/koajs/koa/blob/master/package.json#L42
  - [ ] 👊 bump `express` (via `type-is` library) https://github.com/expressjs/express/blob/master/package.json#L57
  - [ ] 👊 bump `type-is` https://github.com/jshttp/type-is/blob/master/package.json#L13
  - [ ] Update Rails Media Type extensions
  - [ ] Update Rack Contrib Media Type extensions
  - [ ] Update `mime-types` https://github.com/jshttp/mime-types/blob/master/package.json#L17
  - [ ] 👊 bump `node-mime` https://github.com/broofa/node-mime/pull/193/commits/f60fbe2515084350058922d23b3d75db1b488111
  - [ ] 👊 bump `compressible` https://github.com/jshttp/compressible/blob/master/package.json#L19
  - [ ] 👊 bump `accepts` https://github.com/jshttp/accepts/blob/master/package.json#L12


## References
  - [Javascript Media Types on the web. The saga continues](https://github.com/nodejs/node-eps/pull/61#issuecomment-321329296)
  - [Web compatible javascript media types](https://html.spec.whatwg.org/multipage/scripting.html#javascript-mime-type)
  - https://github.com/whatwg/html/issues/558
  - https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/http/mime_types.rb
  - https://rubyplus.com/articles/3231-How-To-Get-A-List-Of-All-Currently-Registered-MIME-Types-With-Rails
  - [WHATWG `fetch`ing a single module algorithm](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script)


## ES6 Modules & Client/Server support

Anyone touching an app in the browser should at least get familiar.
/cc @mrbernnz @btakita @halorium @janz93 @pachonk @robcole @scottmacdowell @kurtcagle @johnrlive
  - http://jsmodules.io
  - https://blog.whatwg.org/js-modules
  - http://2ality.com/2017/05/es-module-specifiers.html
  - (@tmornini  you thought THIS PR was Epic. lol) https://github.com/whatwg/html/pull/443

File: middleware/negotiator.es [hunk 1]
```diff
@@ -36,7 +36,7 @@ module.exports = async (context, next) =>
 
   await next ()
 
-  console.log (extension, resource)
+  //console.log (extension, resource)
```
Comment (tmornini) [inline middleware/negotiator.es]: Not a fan of commented-out code.

File: middleware/route.test [hunk 2]
```diff
@@ -109,7 +109,7 @@ test ('decoding route URI encoded {token} parameters', async t => {
 })
 
 
-test.only ('', async t => {
+test.only ('Testin bitch', async t => {
```
Comment (tmornini) [inline middleware/route.test]: HTTP 406 - not acceptable

File: middleware/route.test [hunk 3]
```diff
@@ -125,7 +125,7 @@ test.only ('', async t => {
   , server =
       (new Server (stack)).serve ``
 
-  , response = await fetch (`${URI}${encoded}`)
+  , response = await fetch (`${URI}${value}`)
```
Comment (tmornini) [inline middleware/route.test]: Why code more-generic? The original name is so much more clear...

Comment (snuggs) [reply middleware/route.test]: @tmornini @brandondees  wrapping this one up just making sure I covered everything asked for. `encoded` in retrospect is better off. "token/value" pairs are the most abstract. The automobile to the skateboard.

#FunFact @tmornini they don't call it HTTP method in the spec they call it `method *token*` ;-) 

capture d ecran 2018-05-09 a 11 37 57

**Comment by snuggs** [comment]
@brandondees @tmornini irony. At the WHATWG/W3C we've come to consensus that the defacto mime for JS is `text/javascript` However, in Ruby, Node, and I believe Go the main mime libraries do not reflect for `text/javascript` (But do wrongly assume `application/javascript` as the old spec from 2003 that we are updating has always lead devs down the wrong mimetype path (pun intended).

So the last remaining tests in suite are failing until this bug/feature gets added.

capture d ecran 2018-06-06 a 14 15 09

**Comment by snuggs** [comment]
@tmornini as per your recommendation to skip for time being. Have created issue #174 to track progress.

I think that's a wrap!!

Please accept.

**Comment by tmornini** [comment]
> @tmornini may need your guidance on the protocol to update golang source code. text/x-javascript is on the MUST NOT USE list as far as the platform is concerned. What should we do to notify them? https://golang.org/src/mime/type.go

https://github.com/golang/go/issues

**Pull request merged**